presto-main/src/main/java/com/facebook/presto/execution/TaskExecutor.java

```diff
@@ -991,4 +1049,155 @@ public ThreadPoolExecutorMBean getProcessorExecutor()
     {
         return executorMBean;
     }
+
+    private static TaskExectorController createTaskExecutorController(TaskManagerConfig config)
+    {
+        requireNonNull(config, "config is null");
+        checkArgument(config.getMinWorkerThreads() <= config.getMaxWorkerThreads(), "minWorkerThreads should be less than maxWorkerThreads");
+
+        if (config.getMinWorkerThreads() == config.getMaxWorkerThreads()) {
+            return new StaticTaskExecutorController(config.getMaxWorkerThreads());
+        }
+
+        // choose conservative parameters for PID (kp, ki, kd) controller as default values,
+        // these parameters (kp, ki, kd) might be exposed to user in the future
+        return new PidTaskExecutorController(
+                config.getTargetCpuUtilization(),
+                config.getMinWorkerThreads(),
+                config.getMaxWorkerThreads(),
+                0.1,
+                0.1,
+                0.01);
+    }
+
+    private class DriverManagerThread
+            implements Runnable
+    {
+        private static final String OS_OBJECT_NAME = "java.lang:type=OperatingSystem";
+        private static final String ATTRIBUTE_NAME = "ProcessCpuTime";
+
+        private final MBeanServer mBeanServer;
+        private ObjectName objectName;
+        private long lastWallNanos;
+        private long lastCpuNanos;
+
+        public DriverManagerThread()
+        {
+            mBeanServer = ManagementFactory.getPlatformMBeanServer();
+            try {
+                objectName = new ObjectName(OS_OBJECT_NAME);
+            }
+            catch (MalformedObjectNameException e) {
+                objectName = null;
+            }
+            Long cpuTime = getProcessCpuTime();
+            if (cpuTime != null) {
+                lastCpuNanos = cpuTime.longValue();
+                lastWallNanos = System.nanoTime();
+            }
+        }
+
+        @Override
+        public void run()
+        {
+            if (objectName == null || taskExectorController instanceof StaticTaskExecutorController) {
+                // Do not periodically update the number of runner threads if:
+                // 1. stat could not be correctly generated, or
+                // 2. the number of runner threads is fixed
+                return;
+            }
+            try (SetThreadName threadName = new SetThreadName("DriverManager")) {
+                while (!closed) {
+                    try {
+                        Thread.sleep(runnerThreadsAdjustmentInterval.toMillis());
```

Comment:
plaese do not do model this as thread with internal sleeping.
Use `ScheduledExecutorService .scheduleWithFixedDelay` instead.